### DISCUSSION TITLE
Tech stack conversation.

### DISCUSSION THREAD

**Discussion by Philip-Clark** [club-40/odin-pantry#14] [Ideas]
Feel free to suggest a tech stack.

Update: This discussion can begin in earnest now that the project has been selected.

**Comment by MarchandMD** [comment]
+1 for a RoR backend

**Comment by CChiu28** [comment]
Recommending React(Vite)/Typescript/Tailwind for frontend

**Reply by Lofty-Brambles** [reply]
I'm sure vite is a cool choice. But we should definitely be adding something more developed. 
[Vite-plugin-ssr](https://vite-plugin-ssr.com/) is definitely a solid choice, not only does it pre-render content for SEO, but it can also take a lot of care about the rest of the issues that next deals with.
I'll lead an update on this once the MVP gets finalized though, as we need some key questions answered.

**Reply by Lofty-Brambles** [reply]
As for tailwind, we are surely using an structural library and not writing off-the-screen classes like a maniac right? Off the top of my head we have two options, Daisy UI and Material UI. Off the top of my head, Daisy offers simple intellisense too, and also is easy enough to work with, so it's the better option for me.

**Reply by justEstif** [reply]
Radix ui has been growing in popularity. [this(not sure it has a name yet)](https://ui.shadcn.com/) combines radix ui with tailwind. Might be work considering. Seems as well.

**Reply by Lofty-Brambles** [reply]
Yea, radixUI is cool enough, but daisy has a ton more options and is much more mature.. Either ways, I'm fine with any.

**Reply by ManorSailor** [reply]
I haven't been keeping up with discussion on Discord & here, a lot has happened in my absence. Anyways. In PR #25, it seems we have moved forward with DaisyUI. I've also come across [Preline](https://preline.co/), a UI library with _template style_  approach. Any thoughts on Preline?
Additionally, the PR is missing [prettier-plugin-tailwindcss](https://github.com/tailwindlabs/prettier-plugin-tailwindcss) for sorting class names in the correct order. Despite using a UI library, we may still want to customize it using tailwind classes. In that scenario, having the classes sorted in the _correct_ order would be beneficial.

**Comment by justEstif** [comment]
Recommending Express + Prisma(SQL) if we aren't going with rails.

**Reply by MarchandMD** [reply]
ooh, yeah, i'm super ok with Express too; would be great to work on a Node project

**Reply by Philip-Clark** [reply]
It hasn't been set in stone yet, but I think the general consensus is that the BE will be Ruby. This is largely because TOP teaches BE first in the Ruby path, and last in the Node path. Given that, more people from both paths can work on more parts if the backend is Ruby.

But definitely keep the ideas flowing. If we go with Node as BE, these comments will be useful.

**Comment by rcamach7** [comment]
Frontend Tech Stack:
- **NextJS (React Framework)**
    - Native routing support
    - Organized project structure for better code understanding
    - Easy external backend integration
    - Will offer NextAuth for easy user authentication
    - Built-in optimizations
    - Cons: Learning curve
    - Alternatives: Vite, Remix
- **Material UI (React Component Library)**
    - Speeds up development with pre-built components
    - Consistent component design
    - Responsive
    - Cons: Learning curve, design constraints
    - Alternatives: TailwindCSS, Bootstrap, ChakraUI
- **State Management: React Context**
    - Easy to set up
    - Cons: re-rendering concerns, boilerplate code for complex state updates
    - Alternatives: Redux (might be overkill for small apps)
- **Testing**: … (I don’t have deep experience here)

**Reply by justEstif** [reply]
I would add [zustand](https://github.com/pmndrs/zustand) for state management consideration. Very simple to setup and learn, less boilerplate than react context.

**Reply by Lofty-Brambles** [reply]
Here's some cons against this stack which's why I don't prefer it.

**1. Vite with vite-plugin-ssr**
NextJS is a cool way of developing react apps, but there's some big flaws. It's new features which are so touted, are still in beta, and any app we make now, will need to be updated to use newer features as soon as the next update drops, and that update is *severely* different. I don't see a point in using it for a fresh app atm. It's highly integrated to running a fully fledged node app with serverless functions which makes us extremely dependant on certain hosts.
Vite is a very stable ecosystem and `vite-plugin-ssr` allows a ton of freedom compared to Next with the way you can run an app as a middleware. It's just as fully fledged and minimal, with obviously, the rapid HMR that vite has. It's SSR capabilities are much better, offers us to tap into server-side routing, AND doesn't stick us with one tool for things like image optimisation like next does. It has in built i18n capabilities can be hosted anywhere, AND is a simple vite plugin, so any vite plugin can be used later.

**Reply by Lofty-Brambles** [reply]
**2. Tailwind with a basic structure framework**
I don't think people will like their app to look boring. An UI library that's as painful to customise as material is not something that's necessary when options like DaisyUI and Ripple exists where you can simply add tailwind-like classes to create a big structure. 
Plus, in general, these look much nicer than Material.

**Reply by rcamach7** [reply]
> Here's some cons against this stack which's why I don't prefer it.
> 
> **1. Vite with vite-plugin-ssr** NextJS is a cool way of developing react apps, but there's some big flaws. It's new features which are so touted, are still in beta, and any app we make now, will need to be updated to use newer features as soon as the next update drops, and that update is _severely_ different. I don't see a point in using it for a fresh app atm. It's highly integrated to running a fully fledged node app with serverless functions which makes us extremely dependant on certain hosts. Vite is a very stable ecosystem and `vite-plugin-ssr` allows a ton of freedom compared to Next with the way you can run an app as a middleware. It's just as fully fledged and minimal, with obviously, the rapid HMR that vite has. It's SSR capabilities are much better, offers us to tap into server-side routing, AND doesn't stick us with one tool for things like image optimisation like next does. It has in built i18n capabilities can be hosted anywhere, AND is a simple vite plugin, so any vite plugin can be used later.

I totally see your point. I think this is where we decide what works best for our needs. Do we go for a super customizable build tool, or a framework with core features built in? Keep in mind, though, more customization can also mean a bigger learning curve and more complexity. Since we're just building a CRUD app, maybe we might not need all that extra customization.

Also, Next.js is pretty solid and lets us adopt new features gradually, so we don't have to mess with any beta stuff if we don't want to. Plus, we're not tied to Vercel for hosting or anything, so we've got options there too.

**Reply by rcamach7** [reply]
> **2. Tailwind with a basic structure framework** I don't think people will like their app to look boring. An UI library that's as painful to customise as material is not something that's necessary when options like DaisyUI and Ripple exists where you can simply add tailwind-like classes to create a big structure. Plus, in general, these look much nicer than Material.

DaisyUI and Ripple are solid alternatives! 💪

**Reply by Lofty-Brambles** [reply]
> I totally see your point. I think this is where we decide what works best for our needs. Do we go for a super customizable build tool, or a framework with core features built in? Keep in mind, though, more customization can also mean a bigger learning curve and more complexity. Since we're just building a CRUD app, maybe we might not need all that extra customization.

Yep, which's __why__ I'm suggesting `vite-plugin-ssr`. Not only is it vite-based, it offers all the goodness of next, like file system routing, and is easy to integrate into a rails backend. Next pushes serverless infrastructure so harshly, that it is a pain (and I don't even know if it's possible) to make a rails backend integrate into it.

**Reply by rcamach7** [reply]
> Yep, which's **why** I'm suggesting `vite-plugin-ssr`. Not only is it vite-based, it offers all the goodness of next, like file system routing, and is easy to integrate into a rails backend. Next pushes serverless infrastructure so harshly, that it is a pain (and I don't even know if it's possible) to make a rails backend integrate into it.

Vite is just a build tool and doesn't offer as much as Next provides natively (this can be a good thing in many situations). Plus, it's been around a bit longer, so there are loads of plugins and libraries to help you out. It also integrates with any backend, regardless what languages or frameworks they're built with.

**Reply by fcasibu** [reply]
Storybook for visual testing and a more streamlined way to test components in isolation (should be enough, it's also super easy to write stories). I'm not totally against testing with jest/rtl or even e2e tests, I'll go with the majority, but personally 🙅‍♂️

**Reply by Lofty-Brambles** [reply]
Did you bother reading `vite-plugin-ssr` docs, or is this just.. assuming what it does? It even offers a neat table to show it's features when compared to Next on what it tries to fix. What is Next other than a glorified build tool with a ton more features?
A lot more people are familiar with some solution that's more vanilla react, which's much easier to use.

**Reply by crespire** [reply]
>     * **Testing**: … (I don’t have deep experience here)

For the Rails side, I would recommend RSpec set up with Faker and focusing on request tests. On the front end side, I am less experienced. For a similiar project, I ended up using Cypress e2e test suites but mocking the entire backend in the suite. I don't recommend this route, as it's very brittle.

**Comment by taladan** [comment]
I have no suggestions as far as a tech stack goes as I don't have a deep enough knowledge about everything that is available.  I am currently in the ruby section of the course so, if we go with ROR, technically I should be able to at least play like I'm helping with the back end stuff.  What is our deadline for deciding?

**Reply by justEstif** [reply]
According to the timeline, April 12. I don't think it is a strict deadline however.

**Reply by Ananth-Adhikarla** [reply]
I think Biff said by mid or end of this week to finalize and close it

**Comment by MarchandMD** [comment]
Do we need to have a conversation RE versions of software? I.E. Rails 5 v rails 6 v Rails 7. Or is it best to stick with the current stable curriculum versions?

**Reply by crespire** [reply]
I think sticking with the stable versions of the tools should be fine. That would mean the latest Rails 7, latest React, and all the libraries that support those versions.

**Comment by crespire** [comment]
A few comments re: tech stack and library choices.

#### Authorization on the API
There are a lot of excellent authorization gems out there, and the product should probably use one of those rather than write something from scratch. I've worked a lot with Pundit, but CanCanCan is also an option. User permissions like being able to delete, edit, update, invite other users, etc to a pantry should be built on top of an authorization system based on these gems. We should probably decide on an option before the project gets underway.


#### Front end state and data management
We should also talk about how we will manage state on the front end application and how it will interact/interface with the backend. There are a lot of React library options for this, some more fully featured than others.
  * Zustand is the "it girl" of the moment, but I have no opinion on it, other than it seems very light and very focused on client side. It would likely be useful to use it in conjunction with React Query.
  * I've personally only used React Query/Tanstack Query, and I used it with vanilla React state management to get what I wanted done. Pairing with Zustand might give us all the functionality we need. It is also possible to pair React Query with React's native state management systems (reducers, `useState`, and `useContext`) to get an acceptable result.
  * Redux with redux toolkit (RTK) also provides a similiar kind of functionality, but it's what I would consider a "full fat" library. It does both client state management through stores, and client data management via RTK. This is a popular library, with lots of resources out there, so even though I think it might be a bit overkill, I would recommend we at least consider it.

#### Front end UI (and component?) library
I'm a big fan of TailwindCSS plus DaisyUI.

Open to thoughts, criticisms and other input.

**Reply by Lofty-Brambles** [reply]
I'm ok with it, but here, the issue is, pre-rendering. Not all things do it, and it's __very__ necessary for a production ready app to provide SEO. Which's near-impossible for an SPA, and that is why I'm pushing `vite-plugin-ssr` as it's very minimal, offers all the sweet features of next like file system routing, and is very easy to integrate as a middleware.

**Reply by crespire** [reply]
My personal opinion is that SSR is not going to be required to build a good product, but I am open to it, as I haven't used it myself.

For me, the familiarity bias is strong, and I think sticking with vanilla React without SSR will enable more folks to participate. SSR doesn't feel like it would solve any problems we would encounter better than React + API, and the cost to introduce it is quite a bit higher in terms of mental model adjustments, complexity around routing, adding vite, vite-test, etc.

**Reply by JoshDevHub** [reply]
Do you have any thoughts about what we go with for Authentication? Devise seems pretty annoying with a separate front end + JWT stuff. I feel like there's probably something we could do with a provider (Auth0 maybe?), but I don't have direct experience with that.

**Reply by Lofty-Brambles** [reply]
`vite-plugin-ssr` is a namesake for a ton of QoL features. You can definitely stick to CSR and only pre-render a few pages as necessary. It has file system routing and opt in features, as necessary.
You can definitely add anything you prefer and use much more vanilla react than you'd use than next.

<https://vite-plugin-ssr.com/render-modes> CSR.
<https://vite-plugin-ssr.com/prerender-config#partial> Partial prerendering
<https://vite-plugin-ssr.com/filesystem-routing> File system routing
<https://vite-plugin-ssr.com/nextjs-comparison> Comparison w/ Next.

Plus, the rest of their documentation is very nice and terse.

**Reply by Lofty-Brambles** [reply]
> Do you have any thoughts about what we go with for Authentication? Devise seems pretty annoying with a separate front end + JWT stuff. I feel like there's probably something we could do with a provider (Auth0 maybe?), but I don't have direct experience with that.

I do understand that Auth0 is very easy to integrate and free up to 8k users, but I've only used it with Node.
Does <https://auth0.com/docs/quickstart/backend/rails> sound handy?
Integrating into react is as simple as a ctx provider: <https://auth0.com/docs/quickstart/spa/react>
@JoshDevHub

**Reply by crespire** [reply]
> Do you have any thoughts about what we go with for Authentication? Devise seems pretty annoying with a separate front end + JWT stuff. I feel like there's probably something we could do with a provider (Auth0 maybe?), but I don't have direct experience with that.

I think auth0 is an excellent option. Devise is tough in this project because we're not using the Rails frontend. My only hesitation is that Auth0 seems to offer authorization at the token level, but I think we should handle authorization in-app via Pundit or CanCanCan, rather than try to tie authorization in with authentication via Auth0.

**Reply by crespire** [reply]
> `vite-plugin-ssr` is a namesake for a ton of QoL features. You can definitely stick to CSR and only pre-render a few pages as necessary. It has file system routing and opt in features, as necessary. You can definitely add anything you prefer and use much more vanilla react than you'd use than next.
> 
> https://vite-plugin-ssr.com/render-modes CSR. https://vite-plugin-ssr.com/prerender-config#partial Partial prerendering https://vite-plugin-ssr.com/filesystem-routing File system routing https://vite-plugin-ssr.com/nextjs-comparison Comparison w/ Next.
> 
> Plus, the rest of their documentation is very nice and terse.

I don't think terse documentation is a boon for a project of this nature. I don't know, I just don't see the net positive here of going with something like `vite-plugin-ssr`. Not a blocker for me but I just have some reservations about using it for the sake of using it. I am happy to leave the decision to the rest of the team.

**Reply by MarchandMD** [reply]
> Do you have any thoughts about what we go with for Authentication? Devise seems pretty annoying with a separate front end + JWT stuff. I feel like there's probably something we could do with a provider (Auth0 maybe?), but I don't have direct experience with that.

Any opinions RE omniauth-google-oauth2 gem? It's a breeze to set up....

**Comment by rcamach7** [comment]
## Frontend Tech Stack (revised)
- **Language**: Javascript (versatile, widely-used) or Typescript (type safety, robust)
- **UI Library**: React (component-based, efficient)
- **Build Tool**: Vite (modern, fast)
- **State Management**: Zustand (lightweight, simple API, no boilerplate) or Redux with RTK (predictable, excellent tooling)
- **Styling Solution**: TailwindCSS (utility-first, rapid development)
- **Testing**: Jest (supports React, part of Odin Curriculum)
- **Authentication**: JWT-based authentication (stateless, scalable)
- **Other**: ReactRouterDOM (popular client-side routing), ESLint (enforces code quality), Prettier (consistent code formatting), EditorConfig (unified editor settings)
- **Worth Considering**: Git Commit Linting (standardized commit messages, improved collaboration)

## Backend Tech Stack
- **Language**: Ruby (dynamic, object-oriented)
- **Framework**: Ruby On Rails (powerful, MVC pattern)
- **Authentication**: JWT-based authentication (stateless, scalable, widely supported)
  - Authorization: Pundit (flexible, easy to use)
- **Database**: PostgreSQL (structured, scalable)
- **Testing**: RSpec (mature, widely used)

**Reply by Lofty-Brambles** [reply]
Commenting on particular points here.

**4.** State Management library: As @crespire mentioned, and I'd vouch too, Zustand as a drop in state management library is very handy over something that's like React Context. It has a much less overhead than Redux, and better optimization with rebuilding components as necessary compared to React Context, and essentially acts as a custom hook itself..

**Reply by Lofty-Brambles** [reply]
**5.** I'd say the debate was pretty much settled on Tailwind. The only other discussion was on what UI component library was to be added to avoid using Tailwind like a true masochist, and the majority of people seemed to agree on DaisyUI as a solid choice. Feel free to differ on this.
No idea why styled components are mentioned. Even the react team has shafted it as a terrible solution.

**Reply by crespire** [reply]
Will also flag on the backend side, we should use an authorization gem. Two good options are Pundit and CanCanCan

**Reply by Lofty-Brambles** [reply]
Clubbing the rest of the points together...
**1.** I thought it was Typescript? 
**6.** Testing, consider discussing on using storybook as a solution. It offers a nice visual interface. The only reason I can see, _not_ to use it is the way it's maintained/bundled, makes it a pain to maintain a project and not be outdated. I, personally, do not see a reason to unit test React components, but some other's may disagree, so I'll leave it to them to test what they find necessary (or atleast point it out)... Cypress offers some neat e2e features, maybe worth exploring.
**8.** Definitely worth it, taking some time to write out a plugin or something to add file system routing over using raw react-router-dom. It's straight up painful. Axios is just unnecessary over native fetch. I don't see the need of it's use in a non-legacy project, unless someone is really used to it, which is not true for a group of people, tbh. Not sure how editorconfig helps over prettier, so I'll leave this up for those who know.
**9.** Conventional commits, atleast during merging is requested in the `CONTRIBUTING.md`.

**Reply by mmxz-mmxz** [reply]
I think there's a strong case to lean in harder to feature/domain driven design over the current paradigm where only components are organized by feature. I also recommend [Storybook](https://storybook.js.org/) as part of our FE stack. The folder tree would look like so:  
  
```
├── .storybook/          # Contains Storybook config for the application.
├── src/                      # Contains all source code for the project.
│   ├── common/              # Contains common components/api/hooks/types/utils/etc. used by application  
│   │   │   ├── components/  # Contains all common components  
│   │   │   ├── api/          # Contains all common api calls
│   │   │   ├── hooks/          # Contains all common hooks  
│   │   │   ├── types/          # Contains all common types  
│   │   │   └──  utils/          # Contains all common utils  
│   ├── modules (or features, etc.)/ # Contains images, fonts, and other static assets that will be used by the application.
│   │   ├── feature1/          # Contains all components/util/api/styles/etc. relevant to feature1
│   │   │   ├── components/  # Contains all components relevant to feature1  
│   │   │   ├── pages/          # Contains all pages relevant to feature1
│   │   │   ├── api/          # Contains all api calls relevant to feature1  
│   │   │   ├── hooks/          # Contains all hooks relevant to feature1  
│   │   │   ├── context/          # Contains all context relevant to feature1  
│   │   │   ├── types/          # Contains all common types relevant to feature1
│   │   │   └──  utils/          # Contains all general utils relevant to feature1  
│   ├── app/              # Contains files (main app container, routing) relevant to top level app
│   │   │   ├── App.tsx          # The main application component that initializes routing/state  
│   │   │   └──  routes.tsx          # Contains routing information for the app
│   └──  index.tsx            # The entry point of the application.
```  
  
I also think it's easier to organize styles, unit tests, etc. at the component level, rather than having separate folders at the root level:  
  
```
├── component/                     
│   ├── styles.(ts|css)             
│   ├── index.stories.tsx  # Component stories in storybook   
│   ├── index.test.(ts|tsx)  # Unit tests for component
│   └── index.tsx 
```  
  
Another suggestion is dropping auth on our end in favor an auth provider - think this is just a lot easier.

**Reply by mmxz-mmxz** [reply]
Also just want to bring up [react-query](https://tanstack.com/query/v3/) again - we should probably consider this if we are doing CSR

**Reply by rcamach7** [reply]
Updated. I added Zustand as state management, and TailWindCss as the styling solution, per feedback. I also added authorization libraries for the backend. @mmxz-mmxz, are you suggesting react-query as an alternative to Zustand, or a data fetching tool? I have heard a lot about it but I haven't used. 

@crespire brought up a great point on leaning on technologies that are most familiar to most Odin members. @Lofty-Brambles 
 Typescript will fall outside of that and will add a big learning curve to the project, which is why I believe Javascript is the right language for this. On the commit linting, it would allow us to force commit messages to follow a certain format, that we can add on the contributing guidelines. For example, we can force commit messages to start with "feat: <message>", "doc: <message>", etc. It's something optional that can keep commit messages clean.

**Reply by Lofty-Brambles** [reply]
As far as conventional commits go, yes, that would. That would also allow us to generate neater changelogs as well as better bug tracking and versioning of the project. Which's why conventional commits was a requested pattern and updated in the `CONTRIBUTING.md`. Furthermore, Typescript was an almost unanimously agreed upon conclusion from over on discord as well as other discussions, including the second comment on this discussion and the number of upvotes on the other comments over at [the other styles post](https://github.com/orgs/club-40/discussions/6)

**Reply by JaeGif** [reply]
> Also just want to bring up [react-query](https://tanstack.com/query/v3/) again - we should probably consider this if we are doing CSR

I’m a big fan of React Query. Small skill floor high skill ceiling compared to Redux, but most importantly, it’s just easy to use.

**Reply by RyanMcEntire** [reply]
@mmxz-mmxz In looking over the history of the project, you seem to have a concept for how the storybook components can be organized. It looks great. If you're interested in taking ownership of this section to plan how you see fit, I'd love to get going and make some progress on this part of the project.

It seems to me like we just need to pick a structure, and to break things down into groups of tasks that can be completed by individuals. Would you be able to take that on? The figma wireframe is complete, so it just needs to be broken down into actionable items.

**Comment by justEstif** [comment]
If vite is the build tool we are using, vitest is also worth considering.

> Jest represents a duplication of complexity. If your app is powered by Vite, having two different pipelines to configure and maintain is not justifiable. With Vitest you get to define the configuration for your dev, build and test environments as a single pipeline, sharing the same plugins and the same vite.config.js. 

> * Vitest offers compatibility with most of the Jest API and ecosystem libraries
> * faster run for your unit tests and a jump in DX thanks to the default watch mode using Vite instant Hot Module Reload (HMR)

I understand we are trying to stick with TOP content, but vitest being jest compatible means we avoid a lot of the jest setup while getting the DX boost and ease of setup.

**Comment by crespire** [comment]
I personally think that as long as webpack is in the curriculum, that's what the project should stick with, and I am not on the vite train.

It's like arguing over whether we should use import maps or sprockets or propshaft on the rails side. What does `create-react-app` spit out without any flags? I feel that's what we should use. I don't understand the move to vite, and I feel that vite is not an appropriate build tool for this project.

[edit]
It was pointed out that the react revamp is moving to vite. While I still think webpack is more appropriate for the project, this is not a blocker for me.

**Comment by rcamach7** [comment]
Here's a poll, to see if we can get close to a tech stack decision

https://qkld8z3tn.supersurvey.com

**Comment by rcamach7** [comment]
## Frontend Tech Stack (Revised)

- **Language**: Typescript (type safety, robust) | **77%** in favor
- **UI Library**: React (component-based, efficient)
- **Build Tool**: Vite (modern, fast) | **40%** in favor
- **State Management**: Zustand (lightweight, simple API, no boilerplate) | **80%** in favor
- **Styling Solution**: TailwindCSS (utility-first, rapid development)
  - **Component Library**: DaisyUI (uses tailwind, rapid development)
- **Testing**: Vitest (native to vite, fast) | **59%** in favor
  - **Development & Testing**: Storybook (view components in isolation, easy documentation)
- **Authentication**: JWT-based authentication (stateless, scalable)
- **Additional Technologies** (only including those selected by > 50% of respondents):
  - ReactRouterDOM (popular client-side routing) | **52%** in favor
  - ESLint (enforces code quality) | **100%** in favor
  - Prettier (consistent code formatting) | **100%** in favor
  - React Query (efficient data fetching) | **80.95%** in favor
  - Git Commit Linting (standardized commit messages, improved collaboration) | **71.43%** in favor


## Backend Tech Stack
- **Language**: Ruby (dynamic, object-oriented)
- **Framework**: Ruby On Rails (powerful, MVC pattern)
- **Authentication**: JWT-based authentication (stateless, scalable, widely supported)
  - Authorization: Pundit (flexible, easy to use)
- **Database**: PostgreSQL (structured, scalable)
- **Testing**: RSpec (mature, widely used)

This is based on the results of the [Poll Survey](https://qkld8z3tn.supersurvey.com/). The poll was open for 24 hours and received 22 responses. If there aren't percentages on a technology listed, then it had previously been decided by previous discussions to use those technologies.
